### ISSUE TITLE
Unexpectedly slow builds on CI

### ISSUE THREAD

**Issue by mcomella** [mozilla-mobile/firefox-tv#2463]
### Vision statement / What / Requirements
Severin has pointed out that CI builds take much longer than local builds: currently, CI builds across all variants with UI tests and downloading dependencies are on the order of an hour while local builds with a single variant and no UI tests and no dependency download are around 5 minutes. This seems incorrect.

I wonder if the Taskcluster teams, other Android teams, or releng have insight or could help us fix these issues.

### Impact
Slow CI builds create the following problems:
- It's difficult to land changes that need to land quickly: e.g. hotfixes, build version bumps on release day, etc.
- There's a long feedback loop when debugging failures on CI, e.g. #2378 
- If we need to wait for CI to finish to merge a change (e.g. version bump before building a release), it takes up mental attention waiting for a long running CI job to finish
- If we need to build work upon other unlanded work, it can be awkward and time consuming to work around (e.g. merge conflicts); if we can get PRs landed faster, this is less of a problem
- Builds will only take longer in the future so we may wish to solve the issues now

### Acceptance criteria
- [ ] TC builds are reasonably fast compared to local builds (I'd expect ~30min tops)

---

### Research
When running [the main PR command](https://searchfox.org/mozilla-mobile/rev/986178329c866526424089f6641117fc9f3ca757/firefox-tv/.taskcluster.yml#42) – `./gradlew -PisPullRequest clean assembleSystem assembleAndroidTest lint checkstyle ktlint pmd detekt test`, it takes:
- Locally, it takes 7m 13s
- In Docker locally, it takes 16m 34s
- On TC ([this run](https://tools.taskcluster.net/groups/LMNa_ixYQaSd-biOMpIUHQ/tasks/LMNa_ixYQaSd-biOMpIUHQ/runs/0/logs/public%2Flogs%2Flive.log)), it takes 27m 48s

Also, in that TC run, it takes ten minutes from "Task Finished" to "Successful task run":
```
[taskcluster 2019-06-25 18:31:36.718Z] === Task Finished ===
[taskcluster 2019-06-25 18:41:32.367Z] Successful task run with exit code: 0 completed in 3473.922 seconds
```

I also wonder what we can parallelize: e.g. the jacocoTestReport step can run in parallel (in a separate TC task?) to the main build, as well as different variants running in separate tasks.

**Comment by mcomella** [comment]
I wonder if the TC instance has low priority compared to other VMs: releng mentioned the new TC set-up would give us higher priority instances, iirc: I wonder if that will help at all. 

I also wonder if the TC instance has low memory available: I increased the amount of memory available to my local Docker and builds went from 16.5m to 12m.

**Comment by mcomella** [comment]
bstack pointed out, "is it intended that it is polling bitbar every 10 minutes? That seems like a long intercal" – specifically for merges to master. We should investigate.

**Comment by mcomella** [comment]
bstack in #taskcluster helped debug this:
- In general, we shouldn't expect slowdown because we're running on ec2
- Our worker types may be too small, which could slow down our builds: we're using the `github-worker` (as defined by [this task](https://tools.taskcluster.net/groups/CGX3kHk4R9SEqvzycpfxug/tasks/CGX3kHk4R9SEqvzycpfxug/details)), which uses the `m3.xlarge` and `r3.xlarge` instance types ([as defined here](https://tools.taskcluster.net/aws-provisioner/github-worker/view)) with a capacity of 2 (can be shared with another task). These seem reasonably sized (via [ec2 docs](https://aws.amazon.com/ec2/previous-generation/)).
  - We used to have a `c3.xlarge` worker type but bstack removed it. It only had 7.5GB of RAM which seems too small for our gradle daemon tasks (which are usually 2 processes per build with 2-4GB each)
- bstack suggested we could also set capacity to 1 or upgrade instances to improve performance

We might be able to improve performance too by reducing RAM use by disabling the gradle daemon in our builds.

**Comment by pocmo** [comment]
A random list of things we did in AC to speed up our builds (may not be complete):

- Different instance types and lifetimes - but that got expensive pretty quickly, see: https://bugzilla.mozilla.org/show_bug.cgi?id=1558601
- Minimizing the size of the Docker image (for situations where we start building with a fresh instance)
- Rebuild the Docker image every now and then. If there are a lot of new dependencies over time (especially new gradle versions and plugins!) then the task will spend a lot of time just downloading things. Caching those things in the new image always has a big impact for us (and also reduces intermittent build failures because of taskcluster being unable to fetch things)
- Parallelize tasks: Instead of having one task doing all the work spawn multiple tasks doing parts of it. As an "app" example see reference browser, e.g.: https://tools.taskcluster.net/groups/Voo6NOHVST6toQNyrao6Ag
- Running lint on a specific task: Lint is one of the slowest tasks and just doing `lintDebug` or `lintRelease` cuts that time in half.
- Experimenting with `parallel` mode. It did speed some things up, but also caused a bunch of intermittents. This may be less a problem for an app project that doesn't have many Gradle modules.
- Reducing the number of build variants (or disabling some)

Things we didn't try yet - but may be in the future:
* Using gradle enterprise or only its build cache
* Using taskcluster's cache nodes to mount a "task local" build cache
* Cache dependencies close to the task (e.g. using a maven mirror/proxy)
* Let taskcluster build the Docker images automatically if needed

**Comment by liuche** [comment]
Labeling this as a P1 bc it affects all of our development work, so if we can speed up builds (mainly for release builds, and somewhat for PRs), this would affect our speed of future work.

**Comment by mcomella** [comment]
bstack is helping us debug and is tracking in this issue: https://bugzilla.mozilla.org/show_bug.cgi?id=1563010

He changed our worker types to modern hardware & reducing the number of concurrent jobs to 1. Since gradle uses so much memory, he was concerned having multiple workers is slowing down our jobs.

**Comment by mcomella** [comment]
I checked out git tag `v3.4` and builds back then took ~8 minutes. I wonder what changed. Maybe we've included a much larger dependency (Rx?)?
